File: examples/phone-rpc-client.js
#!/usr/bin/env node
/**
 * Phone Number RPC Client Example (Node.js)
 * 
 * This example demonstrates how to interact with the fritz-callmonitor2mqtt
 * phone number RPC API using Node.js and the mqtt library.
 * 
 * Prerequisites:
 *     npm install mqtt
 * 
 * Usage:
 *     node phone-rpc-client.js --help
 *     node phone-rpc-client.js set +1234567890 "John Doe"
 *     node phone-rpc-client.js get +1234567890
 *     node phone-rpc-client.js list
 *     node phone-rpc-client.js search "John"
 *     node phone-rpc-client.js delete +1234567890
 */

const mqtt = require('mqtt');
const { v4: uuidv4 } = require('crypto');

class PhoneNumberRPCClient {
    constructor(options = {}) {
        this.brokerUrl = options.brokerUrl || 'mqtt://localhost:1883';
        this.topicPrefix = options.topicPrefix || 'fritz/callmonitor';
        this.username = options.username;
        this.password = options.password;
        
        this.client = null;
        this.responses = new Map();
        this.connected = false;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            const connectOptions = {
                clientId: `phone-rpc-client-${Date.now()}`,
                clean: true
            };

            if (this.username && this.password) {
                connectOptions.username = this.username;
                connectOptions.password = this.password;
            }

            this.client = mqtt.connect(this.brokerUrl, connectOptions);

            this.client.on('connect', () => {
                this.connected = true;
                console.log(`✅ Connected to MQTT broker ${this.brokerUrl}`);
                
                // Subscribe to response topic
                const responseTopic = `${this.topicPrefix}/phone_number/response`;
                this.client.subscribe(responseTopic, { qos: 1 }, (err) => {
                    if (err) {
                        console.error(`❌ Failed to subscribe to ${responseTopic}:`, err);
                        reject(err);
                    } else {
                        console.log(`📡 Subscribed to ${responseTopic}`);
                        resolve();
                    }
                });
            });

            this.client.on('error', (err) => {
                console.error('❌ MQTT connection error:', err);
                reject(err);
            });

            this.client.on('message', (topic, message) => {
                try {
                    const response = JSON.parse(message.toString());
                    const requestId = response.id;
                    if (requestId) {
                        this.responses.set(requestId, response);
                    }
                } catch (err) {
                    console.error('❌ Failed to parse MQTT response:', err);
                }
            });

            this.client.on('close', () => {
                this.connected = false;
                console.log('📡 MQTT connection closed');
            });

            // Set timeout for connection
            setTimeout(() => {
                if (!this.connected) {
                    reject(new Error('Connection timeout'));
                }
            }, 10000);
        });
    }

    disconnect() {
        if (this.client && this.connected) {
            this.client.end(true);
        }
    }

    async sendRPCRequest(method, options = {}) {
        if (!this.connected) {
            throw new Error('Not connected to MQTT broker');
        }

        const requestId = uuidv4();
        const request = {
            id: requestId,
            method: method,
            timestamp: new Date().toISOString(),
            ...options
        };

        // Publish request
        const requestTopic = `${this.topicPrefix}/phone_number/request`;
        const payload = JSON.stringify(request);

        console.log(`📤 Sending ${method} request...`);
        
        return new Promise((resolve, reject) => {
            this.client.publish(requestTopic, payload, { qos: 1 }, (err) => {
                if (err) {
                    reject(new Error(`Failed to publish request: ${err.message}`));
                    return;
                }

                // Wait for response
                const timeout = setTimeout(() => {
                    this.responses.delete(requestId);
                    reject(new Error('Timeout waiting for response'));
                }, 10000);

                const checkResponse = () => {
                    if (this.responses.has(requestId)) {
                        const response = this.responses.get(requestId);
                        this.responses.delete(requestId);
                        clearTimeout(timeout);
                        resolve(response);
                    } else {
                        setTimeout(checkResponse, 100);
                    }
                };

                checkResponse();
            });
        });
    }

    async setPhoneNumber(phoneNumber, name) {
        try {
            const response = await this.sendRPCRequest('set', {
                phone_number: phoneNumber,
                name: name
            });

            if (response.success) {
                const pn = response.phone_number;
                console.log(`✅ Set ${pn.phone_number} → ${pn.name}`);
                return true;
            } else {
                console.error(`❌ Failed to set phone number: ${response.error}`);
                return false;
            }
        } catch (err) {
            console.error(`❌ Error: ${err.message}`);
            return false;
        }
    }

    async getPhoneNumber(phoneNumber) {
        try {
            const response = await this.sendRPCRequest('get', {
                phone_number: phoneNumber
            });

            if (response.success) {
                const pn = response.phone_number;
                console.log(`✅ Found: ${pn.phone_number} → ${pn.name}`);
                if (pn.created_at) console.log(`   Created: ${pn.created_at}`);
                if (pn.updated_at) console.log(`   Updated: ${pn.updated_at}`);
                return pn;
            } else {
                console.error(`❌ Phone number not found: ${response.error}`);
                return null;
            }
        } catch (err) {
            console.error(`❌ Error: ${err.message}`);
            return null;
        }
    }

    async deletePhoneNumber(phoneNumber) {
        try {
            const response = await this.sendRPCRequest('delete', {
                phone_number: phoneNumber
            });

            if (response.success) {
                console.log(`✅ Deleted ${phoneNumber}`);
                return true;
            } else {
                console.error(`❌ Failed to delete phone number: ${response.error}`);
                return false;
            }
        } catch (err) {
            console.error(`❌ Error: ${err.message}`);
            return false;
        }
    }

    async listPhoneNumbers(limit = 100) {
        try {
            const response = await this.sendRPCRequest('list', { limit });

            if (response.success) {
                const phoneNumbers = response.phone_numbers || [];
                const count = response.count || 0;

                console.log(`✅ Found ${count} phone numbers:`);
                phoneNumbers.forEach((pn, index) => {
                    const name = pn.name || '(no name)';
                    console.log(`   ${(index + 1).toString().padStart(2)}. ${pn.phone_number} → ${name}`);
                });

                return phoneNumbers;
            } else {
                console.error(`❌ Failed to list phone numbers: ${response.error}`);
                return null;
            }
        } catch (err) {
            console.error(`❌ Error: ${err.message}`);
            return null;
        }
    }

    async searchPhoneNumbers(pattern, limit = 100) {
        try {
            const response = await this.sendRPCRequest('search', { pattern, limit });

            if (response.success) {
                const phoneNumbers = response.phone_numbers || [];
                const count = response.count || 0;

                console.log(`✅ Found ${count} phone numbers matching '${pattern}':`);
                phoneNumbers.forEach((pn, index) => {
                    const name = pn.name || '(no name)';
                    console.log(`   ${(index + 1).toString().padStart(2)}. ${pn.phone_number} → ${name}`);
                });

                return phoneNumbers;
            } else {
                console.error(`❌ Failed to search phone numbers: ${response.error}`);
                return null;
            }
        } catch (err) {
            console.error(`❌ Error: ${err.message}`);
            return null;
        }
    }
}

// CLI handling
async function main() {
    const args = process.argv.slice(2);
    
    if (args.length === 0 || args[0] === '--help') {
        console.log(`
Phone Number RPC Client (Node.js)

Usage:
    node phone-rpc-client.js <command> [args...]

Commands:
    set <phone_number> <name>     Set phone number with name
    get <phone_number>            Get phone number information  
    delete <phone_number>         Delete phone number
    list [--limit N]              List all phone numbers
    search <pattern> [--limit N]  Search phone numbers by name

Options:
    --broker <url>                MQTT broker URL (default: mqtt://localhost:1883)
    --prefix <prefix>             MQTT topic prefix (default: fritz/callmonitor)
    --username <username>         MQTT username
    --password <password>         MQTT password
    --limit <N>                   Limit results (default: 100)

Examples:
    node phone-rpc-client.js set +1234567890 "John Doe"
    node phone-rpc-client.js get +1234567890
    node phone-rpc-client.js list
    node phone-rpc-client.js search "John"
    node phone-rpc-client.js delete +1234567890
        `);
        return;
    }

    // Parse command line options
    const options = {
        brokerUrl: 'mqtt://localhost:1883',
        topicPrefix: 'fritz/callmonitor',
        limit: 100
    };

    let command = null;
    let commandArgs = [];

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg.startsWith('--')) {
            const option = arg.substring(2);
            const value = args[i + 1];
            if (value && !value.startsWith('--')) {
                options[option] = value;
                i++; // Skip value
            }
        } else if (!command) {
            command = arg;
        } else {
            commandArgs.push(arg);
        }
    }

    if (!command) {
        console.error('❌ No command specified');
        return;
    }

    const client = new PhoneNumberRPCClient(options);

    try {
        await client.connect();

        switch (command) {
            case 'set':
                if (commandArgs.length < 2) {
                    console.error('❌ Usage: set <phone_number> <name>');
                    return;
                }
                await client.setPhoneNumber(commandArgs[0], commandArgs[1]);
                break;

            case 'get':
                if (commandArgs.length < 1) {
                    console.error('❌ Usage: get <phone_number>');
                    return;
                }
                await client.getPhoneNumber(commandArgs[0]);
                break;

            case 'delete':
                if (commandArgs.length < 1) {
                    console.error('❌ Usage: delete <phone_number>');
                    return;
                }
                await client.deletePhoneNumber(commandArgs[0]);
                break;

            case 'list':
                await client.listPhoneNumbers(options.limit);
                break;

            case 'search':
                if (commandArgs.length < 1) {
                    console.error('❌ Usage: search <pattern>');
                    return;
                }
                await client.searchPhoneNumbers(commandArgs[0], options.limit);
                break;

            default:
                console.error(`❌ Unknown command: ${command}`);
        }

    } catch (err) {
        console.error('❌ Error:', err.message);
    } finally {
        client.disconnect();
        // Give some time for graceful disconnect
        setTimeout(() => process.exit(0), 500);
    }
}

// Handle uncaught errors
process.on('uncaughtException', (err) => {
    console.error('❌ Uncaught Exception:', err);
    process.exit(1);
});

process.on('unhandledRejection', (reason, promise) => {
    console.error('❌ Unhandled Rejection at:', promise, 'reason:', reason);
    process.exit(1);
});

if (require.main === module) {
    main();
}

module.exports = PhoneNumberRPCClient;

File: examples/phone-rpc-client.py
#!/usr/bin/env python3
"""
Phone Number RPC Client Example

This example demonstrates how to interact with the fritz-callmonitor2mqtt
phone number RPC API using Python and the paho-mqtt library.

Prerequisites:
    pip install paho-mqtt

Usage:
    python3 phone-rpc-client.py --help
    python3 phone-rpc-client.py set +1234567890 "John Doe"
    python3 phone-rpc-client.py get +1234567890
    python3 phone-rpc-client.py list
    python3 phone-rpc-client.py search "John"
    python3 phone-rpc-client.py delete +1234567890
"""

import json
import time
import uuid
import argparse
import sys
from datetime import datetime
from typing import Optional, Dict, Any

try:
    import paho.mqtt.client as mqtt
except ImportError:
    print("Error: paho-mqtt not installed. Run: pip install paho-mqtt")
    sys.exit(1)

class PhoneNumberRPCClient:
    def __init__(self, broker_host: str = "localhost", broker_port: int = 1883, 
                 topic_prefix: str = "fritz/callmonitor", username: str = None, 
                 password: str = None):
        self.broker_host = broker_host
        self.broker_port = broker_port
        self.topic_prefix = topic_prefix
        self.username = username
        self.password = password
        
        self.client = mqtt.Client()
        self.responses = {}
        self.connected = False
        
        # Setup callbacks
        self.client.on_connect = self._on_connect
        self.client.on_message = self._on_message
        self.client.on_disconnect = self._on_disconnect
        
        # Setup authentication if provided
        if username and password:
            self.client.username_pw_set(username, password)
    
    def _on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            self.connected = True
            print(f"✅ Connected to MQTT broker {self.broker_host}:{self.broker_port}")
            # Subscribe to response topic
            response_topic = f"{self.topic_prefix}/phone_number/response"
            client.subscribe(response_topic, qos=1)
            print(f"📡 Subscribed to {response_topic}")
        else:
            print(f"❌ Failed to connect to MQTT broker. Return code: {rc}")
    
    def _on_disconnect(self, client, userdata, rc):
        self.connected = False
        if rc != 0:
            print("⚠️ Unexpected MQTT disconnection")
    
    def _on_message(self, client, userdata, msg):
        try:
            response = json.loads(msg.payload.decode())
            request_id = response.get('id')
            if request_id:
                self.responses[request_id] = response
        except json.JSONDecodeError as e:
            print(f"❌ Failed to parse MQTT response: {e}")
    
    def connect(self) -> bool:
        """Connect to MQTT broker"""
        try:
            self.client.connect(self.broker_host, self.broker_port, 60)
            self.client.loop_start()
            
            # Wait for connection
            timeout = 10
            start_time = time.time()
            while not self.connected and (time.time() - start_time) < timeout:
                time.sleep(0.1)
            
            return self.connected
        except Exception as e:
            print(f"❌ Connection failed: {e}")
            return False
    
    def disconnect(self):
        """Disconnect from MQTT broker"""
        if self.connected:
            self.client.loop_stop()
            self.client.disconnect()
    
    def _send_rpc_request(self, method: str, **kwargs) -> Optional[Dict[str, Any]]:
        """Send RPC request and wait for response"""
        if not self.connected:
            print("❌ Not connected to MQTT broker")
            return None
        
        request_id = str(uuid.uuid4())
        request = {
            "id": request_id,
            "method": method,
            "timestamp": datetime.utcnow().isoformat() + "Z",
            **kwargs
        }
        
        # Publish request
        request_topic = f"{self.topic_prefix}/phone_number/request"
        payload = json.dumps(request)
        
        print(f"📤 Sending {method} request...")
        result = self.client.publish(request_topic, payload, qos=1)
        
        if result.rc != mqtt.MQTT_ERR_SUCCESS:
            print(f"❌ Failed to publish request: {result.rc}")
            return None
        
        # Wait for response
        timeout = 10
        start_time = time.time()
        while request_id not in self.responses and (time.time() - start_time) < timeout:
            time.sleep(0.1)
        
        if request_id in self.responses:
            return self.responses.pop(request_id)
        else:
            print("⏰ Timeout waiting for response")
            return None
    
    def set_phone_number(self, phone_number: str, name: str) -> bool:
        """Set/update a phone number with name"""
        response = self._send_rpc_request("set", phone_number=phone_number, name=name)
        
        if response and response.get('success'):
            pn = response.get('phone_number', {})
            print(f"✅ Set {pn.get('phone_number')} → {pn.get('name')}")
            return True
        else:
            error = response.get('error', 'Unknown error') if response else 'No response'
            print(f"❌ Failed to set phone number: {error}")
            return False
    
    def get_phone_number(self, phone_number: str) -> Optional[Dict[str, Any]]:
        """Get information for a phone number"""
        response = self._send_rpc_request("get", phone_number=phone_number)
        
        if response and response.get('success'):
            pn = response.get('phone_number', {})
            print(f"✅ Found: {pn.get('phone_number')} → {pn.get('name')}")
            if pn.get('created_at'):
                print(f"   Created: {pn.get('created_at')}")
            if pn.get('updated_at'):
                print(f"   Updated: {pn.get('updated_at')}")
            return pn
        else:
            error = response.get('error', 'Unknown error') if response else 'No response'
            print(f"❌ Phone number not found: {error}")
            return None
    
    def delete_phone_number(self, phone_number: str) -> bool:
        """Delete a phone number"""
        response = self._send_rpc_request("delete", phone_number=phone_number)
        
        if response and response.get('success'):
            print(f"✅ Deleted {phone_number}")
            return True
        else:
            error = response.get('error', 'Unknown error') if response else 'No response'
            print(f"❌ Failed to delete phone number: {error}")
            return False
    
    def list_phone_numbers(self, limit: int = 100) -> Optional[list]:
        """List all phone numbers"""
        response = self._send_rpc_request("list", limit=limit)
        
        if response and response.get('success'):
            phone_numbers = response.get('phone_numbers', [])
            count = response.get('count', 0)
            
            print(f"✅ Found {count} phone numbers:")
            for i, pn in enumerate(phone_numbers, 1):
                name = pn.get('name') or '(no name)'
                print(f"   {i:2d}. {pn.get('phone_number')} → {name}")
            
            return phone_numbers
        else:
            error = response.get('error', 'Unknown error') if response else 'No response'
            print(f"❌ Failed to list phone numbers: {error}")
            return None
    
    def search_phone_numbers(self, pattern: str, limit: int = 100) -> Optional[list]:
        """Search phone numbers by name pattern"""
        response = self._send_rpc_request("search", pattern=pattern, limit=limit)
        
        if response and response.get('success'):
            phone_numbers = response.get('phone_numbers', [])
            count = response.get('count', 0)
            
            print(f"✅ Found {count} phone numbers matching '{pattern}':")
            for i, pn in enumerate(phone_numbers, 1):
                name = pn.get('name') or '(no name)'
                print(f"   {i:2d}. {pn.get('phone_number')} → {name}")
            
            return phone_numbers
        else:
            error = response.get('error', 'Unknown error') if response else 'No response'
            print(f"❌ Failed to search phone numbers: {error}")
            return None

def main():
    parser = argparse.ArgumentParser(description='Phone Number RPC Client')
    parser.add_argument('--broker', default='localhost', help='MQTT broker host')
    parser.add_argument('--port', type=int, default=1883, help='MQTT broker port')
    parser.add_argument('--prefix', default='fritz/callmonitor', help='MQTT topic prefix')
    parser.add_argument('--username', help='MQTT username')
    parser.add_argument('--password', help='MQTT password')
    
    subparsers = parser.add_subparsers(dest='command', help='Available commands')
    
    # Set command
    set_parser = subparsers.add_parser('set', help='Set phone number with name')
    set_parser.add_argument('phone_number', help='Phone number (e.g., +1234567890)')
    set_parser.add_argument('name', help='Contact name (e.g., "John Doe")')
    
    # Get command
    get_parser = subparsers.add_parser('get', help='Get phone number information')
    get_parser.add_argument('phone_number', help='Phone number to lookup')
    
    # Delete command
    delete_parser = subparsers.add_parser('delete', help='Delete phone number')
    delete_parser.add_argument('phone_number', help='Phone number to delete')
    
    # List command
    list_parser = subparsers.add_parser('list', help='List all phone numbers')
    list_parser.add_argument('--limit', type=int, default=100, help='Maximum results')
    
    # Search command
    search_parser = subparsers.add_parser('search', help='Search phone numbers by name')
    search_parser.add_argument('pattern', help='Search pattern')
    search_parser.add_argument('--limit', type=int, default=100, help='Maximum results')
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    # Create and connect client
    client = PhoneNumberRPCClient(
        broker_host=args.broker,
        broker_port=args.port,
        topic_prefix=args.prefix,
        username=args.username,
        password=args.password
    )
    
    if not client.connect():
        print("❌ Failed to connect to MQTT broker")
        return
    
    try:
        # Execute command
        if args.command == 'set':
            client.set_phone_number(args.phone_number, args.name)
        elif args.command == 'get':
            client.get_phone_number(args.phone_number)
        elif args.command == 'delete':
            client.delete_phone_number(args.phone_number)
        elif args.command == 'list':
            client.list_phone_numbers(args.limit)
        elif args.command == 'search':
            client.search_phone_numbers(args.pattern, args.limit)
    
    finally:
        client.disconnect()

if __name__ == '__main__':
    main()
